x/pairing/keeper/epoch_cu.go

```diff
@@ -0,0 +1,273 @@
+package keeper
+
+import (
+	"strconv"
+	"strings"
+
+	"github.com/cosmos/cosmos-sdk/store/cachekv"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lavanet/lava/utils"
+	"github.com/lavanet/lava/x/pairing/types"
+)
+
+// UniqueEpochSession is used to detect double spend attacks
+// It's kept in a epoch-prefixed store with a unique index: provider, project ID, chain ID and session ID
+//
+// ProviderEpochCu is used to track the CU of a specific provider in a specific epoch for unresponsiveness
+// It's kept in a epoch-prefixed store with a unique index: provider address
+//
+// ProviderConsumerEpochCu is used to track the CU between a specific provider and
+// consumer in a specific epoch for payments
+// It's kept in a epoch-prefixed store with a unique index: provider and project ID
+
+/* ########## UniqueEpochSession ############ */
+
+// SetUniqueEpochSession sets a UniqueEpochSession in the store
+func (k Keeper) SetUniqueEpochSession(ctx sdk.Context, epoch uint64, provider string, project string, chainID string, sessionID uint64) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.UniqueEpochSessionKeyPrefix())
+	store.Set(types.UniqueEpochSessionKey(epoch, provider, chainID, project, sessionID), []byte{0})
+}
+
+// IsUniqueEpochSessionExists checks if a UniqueEpochSession exists
+func (k Keeper) IsUniqueEpochSessionExists(ctx sdk.Context, epoch uint64, provider string, project string, chainID string, sessionID uint64) bool {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.UniqueEpochSessionKeyPrefix())
+	b := store.Get(types.UniqueEpochSessionKey(epoch, provider, chainID, project, sessionID))
+	return b != nil
+}
+
+// RemoveAllUniqueEpochSession removes all the UniqueEpochSession objects from the store for a specific epoch
+func (k Keeper) RemoveAllUniqueEpochSession(ctx sdk.Context, epoch uint64) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.UniqueEpochSessionKeyPrefix())
+	iterator := sdk.KVStorePrefixIterator(store, []byte(strconv.FormatUint(epoch, 10)))
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
+		store.Delete(iterator.Key())
+	}
+}
+
+// GetAllUniqueEpochSessionForEpoch gets all the UniqueEpochSession objects from the store for a specific epoch
+func (k Keeper) GetAllUniqueEpochSessionForEpoch(ctx sdk.Context, epoch uint64) []string {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.UniqueEpochSessionKeyPrefix())
+
+	iterator := sdk.KVStorePrefixIterator(store, []byte(strconv.FormatUint(epoch, 10))) // Get an iterator with no prefix to iterate over all keys
+	defer iterator.Close()
+
+	var keys []string
+	for ; iterator.Valid(); iterator.Next() {
+		key := iterator.Key()
+		// Remove the prefix to get the actual UniqueEpochSession key
+		uniqueEpochSessionKey := strings.TrimPrefix(string(key), string(types.UniqueEpochSessionKeyPrefix()))
```

Comment:
this trimming is not needed

Reply:
done. see https://github.com/lavanet/lava/commit/51c70a28857299d4dab82b8907b750d1967f2bf6